Userland/Applications/Mail/MailWidget.cpp

```diff
@@ -0,0 +1,480 @@
+        GUI::MessageBox::show_error(window(), String::formatted("Failed to retrieve mailboxes. The server says: '{}'", response.response_text()));
+        return false;
+    }
+
+    auto& list_items = response.data().list_items();
+
+    m_account_holder = AccountHolder::create();
+    m_account_holder->add_account_with_name_and_mailboxes(username, move(list_items));
+
+    m_mailbox_list->set_model(m_account_holder->mailbox_tree_model());
+    m_mailbox_list->expand_tree();
+
+    return true;
+}
+
+void MailWidget::on_window_close()
+{
+    auto response = move(m_imap_client->send_simple_command(IMAP::CommandType::Logout)->await().release_value().get<IMAP::SolidResponse>());
+    VERIFY(response.status() == IMAP::ResponseStatus::OK);
+
+    m_imap_client->close();
+}
+
+IMAP::MultiPartBodyStructureData const* MailWidget::look_for_alternative_body_structure(IMAP::MultiPartBodyStructureData const& current_body_structure, Vector<int>& position_stack) const
+{
+    if (current_body_structure.media_type.equals_ignoring_case("ALTERNATIVE"))
+        return &current_body_structure;
+
+    int structure_index = 1;
+
+    for (auto& structure : current_body_structure.bodies) {
+        if (structure->data().has<IMAP::BodyStructureData>()) {
+            ++structure_index;
+            continue;
+        }
+
+        position_stack.append(structure_index);
+        auto* potential_alternative_structure = look_for_alternative_body_structure(structure->data().get<IMAP::MultiPartBodyStructureData>(), position_stack);
+
+        if (potential_alternative_structure)
+            return potential_alternative_structure;
+
+        position_stack.take_last();
+        ++structure_index;
+    }
+
+    return nullptr;
+}
+
+Vector<MailWidget::Alternative> MailWidget::get_alternatives(IMAP::MultiPartBodyStructureData const& multi_part_body_structure_data) const
+{
+    Vector<int> position_stack;
+
+    auto* alternative_body_structure = look_for_alternative_body_structure(multi_part_body_structure_data, position_stack);
+    if (!alternative_body_structure)
+        return {};
+
+    Vector<MailWidget::Alternative> alternatives;
+    alternatives.ensure_capacity(alternative_body_structure->bodies.size());
+
+    int alternative_index = 1;
+    for (auto& alternative_body : alternative_body_structure->bodies) {
+        VERIFY(alternative_body->data().has<IMAP::BodyStructureData>());
+
+        position_stack.append(alternative_index);
+
+        MailWidget::Alternative alternative = {
+            .body_structure = alternative_body->data().get<IMAP::BodyStructureData>(),
+            .position = position_stack,
+        };
+        alternatives.append(alternative);
+
+        position_stack.take_last();
+        ++alternative_index;
+    }
+
+    return alternatives;
+}
+
+bool MailWidget::is_supported_alternative(Alternative const& alternative) const
+{
+    auto lowercase_type = alternative.body_structure.type.to_lowercase();
```

Comment:
At least the type should use `equals_ignoring_case` instead of creating a lowercased `String`. With subtype, it's a bit harder, although since there are only two possibilities, you could also use `equals_ignoring_case`, just twice.